File: integration/unity/Assets/Scripts/behaviac/runtime/BehaviorTree/Nodes/Composites/Referencebehavior.cs
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Tencent is pleased to support the open source community by making behaviac available.
//
// Copyright (C) 2015-2017 THL A29 Limited, a Tencent company. All rights reserved.
//
// Licensed under the BSD 3-Clause License (the "License"); you may not use this file except in compliance with
// the License. You may obtain a copy of the License at http://opensource.org/licenses/BSD-3-Clause
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is
// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and limitations under the License.
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////

using System;
using System.Collections.Generic;

namespace behaviac
{
    public class ReferencedBehavior : BehaviorNode
    {
#if BEHAVIAC_USE_HTN
        public override bool decompose(BehaviorNode node, PlannerTaskComplex seqTask, int depth, Planner planner)
        {
            ReferencedBehavior taskSubTree = (ReferencedBehavior)node;
            bool bOk = false;
            Debug.Check(taskSubTree != null);
            int depth2 = planner.GetAgent().Variables.Depth;
            using(AgentState currentState = planner.GetAgent().Variables.Push(false))
            {
                Agent pAgent = planner.GetAgent();

                string szTreePath = taskSubTree.GetReferencedTree(pAgent);
                BehaviorTreeTask subTreeTask = Workspace.Instance.CreateBehaviorTreeTask(szTreePath);

                taskSubTree.SetTaskParams(pAgent, subTreeTask);

                Task task = taskSubTree.RootTaskNode(planner.GetAgent());

                if (task != null)
                {
                    planner.LogPlanReferenceTreeEnter(planner.GetAgent(), taskSubTree);
                    //task.Parent.InstantiatePars(this.LocalVars);

                    BehaviorTreeTask oldCurrentTreeTask = pAgent.ExcutingTreeTask;
                    pAgent.ExcutingTreeTask = subTreeTask;
                    PlannerTask childTask = planner.decomposeNode(task, depth);
                    pAgent.ExcutingTreeTask = oldCurrentTreeTask;

                    if (childTask != null)
                    {
                        //taskSubTree.SetTaskParams(planner.GetAgent(), childTask);
                        PlannerTaskReference subTreeRef = (PlannerTaskReference)seqTask;

                        subTreeRef.SubTreeTask = subTreeTask;
                        seqTask.AddChild(childTask);
                        bOk = true;
                    }

                    //task.Parent.UnInstantiatePars(this.LocalVars);
                    planner.LogPlanReferenceTreeExit(planner.GetAgent(), taskSubTree);
                    Debug.Check(true);
                }
            }

            Debug.Check(planner.GetAgent().Variables.Depth == depth2);
            return bOk;
        }
#endif//

        protected override void load(int version, string agentType, List<property_t> properties)
        {
            base.load(version, agentType, properties);

            for (int i = 0; i < properties.Count; ++i)
            {
                property_t p = properties[i];

                if (p.name == "ReferenceBehavior")
                {
                    int pParenthesis = p.value.IndexOf('(');

                    if (pParenthesis == -1)
                    {
                        this.m_referencedBehaviorPath = AgentMeta.ParseProperty(p.value);
                    }
                    else
                    {
                        this.m_referencedBehaviorPath = AgentMeta.ParseMethod(p.value);
                    }

                    string szTreePath = this.GetReferencedTree(null);

                    //conservatively make it true
                    bool bHasEvents = true;

                    if (!string.IsNullOrEmpty(szTreePath))
                    {
                        if (Config.PreloadBehaviors)
                        {
                            BehaviorTree behaviorTree = Workspace.Instance.LoadBehaviorTree(szTreePath);
                            Debug.Check(behaviorTree != null);

                            if (behaviorTree != null)
                            {
                                bHasEvents = behaviorTree.HasEvents();
                            }
                        }

                        this.m_bHasEvents |= bHasEvents;
                    }
                }
                else if (p.name == "Task")
                {
                    this.m_taskMethod = AgentMeta.ParseMethod(p.value);
                }
            }
        }

        public override void Attach(BehaviorNode pAttachment, bool bIsPrecondition, bool bIsEffector, bool bIsTransition)
        {
            if (bIsTransition)
            {
                Debug.Check(!bIsEffector && !bIsPrecondition);

                if (this.m_transitions == null)
                {
                    this.m_transitions = new List<Transition>();
                }

                Transition pTransition = pAttachment as Transition;
                Debug.Check(pTransition != null);
                this.m_transitions.Add(pTransition);

                return;
            }

            Debug.Check(bIsTransition == false);
            base.Attach(pAttachment, bIsPrecondition, bIsEffector, bIsTransition);
        }

        public override bool IsValid(Agent pAgent, BehaviorTask pTask)
        {
            if (!(pTask.GetNode() is ReferencedBehavior))
            {
                return false;
            }

            return base.IsValid(pAgent, pTask);
        }

        protected override BehaviorTask createTask()
        {
            ReferencedBehaviorTask pTask = new ReferencedBehaviorTask();

            return pTask;
        }

        public virtual string GetReferencedTree(Agent pAgent)
        {
            if (this.m_referencedBehaviorPath != null)
            {
                Debug.Check(this.m_referencedBehaviorPath is CInstanceMember<string>);
                return ((CInstanceMember<string>)this.m_referencedBehaviorPath).GetValue(pAgent);
            }

            return string.Empty;
        }

        public void SetTaskParams(Agent agent, BehaviorTreeTask treeTask)
        {
            if (this.m_taskMethod != null)
            {
                this.m_taskMethod.SetTaskParams(agent, treeTask);
            }
        }

        protected List<Transition> m_transitions;
        protected IInstanceMember m_referencedBehaviorPath;
        protected IMethod m_taskMethod;

        private Task m_taskNode;

        public Task RootTaskNode(Agent pAgent)
        {
            if (this.m_taskNode == null)
            {
                string szTreePath = this.GetReferencedTree(pAgent);
                BehaviorTree bt = Workspace.Instance.LoadBehaviorTree(szTreePath);

                if (bt != null && bt.GetChildrenCount() == 1)
                {
                    BehaviorNode root = bt.GetChild(0);
                    this.m_taskNode = root as Task;
                }
            }

            return this.m_taskNode;
        }

        public class ReferencedBehaviorTask : SingeChildTask
        {
#if BEHAVIAC_USE_HTN
            private AgentState currentState;
#endif//
            //~ReferencedBehaviorTask()
            //{
            //    Workspace.Instance.DestroyBehaviorTreeTask(this.m_subTree, null);
            //    this.m_subTree = null;
            //}

            protected override bool CheckPreconditions(Agent pAgent, bool bIsAlive)
            {
#if BEHAVIAC_USE_HTN
                this.currentState = pAgent.Variables.Push(false);
                Debug.Check(currentState != null);
#endif//

                bool bOk = base.CheckPreconditions(pAgent, bIsAlive);
#if BEHAVIAC_USE_HTN

                if (!bOk)
                {
                    this.currentState.Pop();
                    this.currentState = null;
                }

#endif//
                return bOk;
            }

            int m_nextStateId = -1;
            public override int GetNextStateId()
            {
                return m_nextStateId;
            }

            BehaviorTreeTask m_subTree = null;

            public override bool onevent(Agent pAgent, string eventName, Dictionary<uint, IInstantiatedVariable> eventPrams)
            {
                if (this.m_status == EBTStatus.BT_RUNNING && this.m_node.HasEvents())
                {
                    Debug.Check(this.m_subTree != null);

                    if (!this.m_subTree.onevent(pAgent, eventName, eventPrams))
                    {
                        return false;
                    }
                }

                return true;
            }

            protected override bool onenter(Agent pAgent)
            {
                ReferencedBehavior pNode = this.GetNode() as ReferencedBehavior;
                Debug.Check(pNode != null);

                if (pNode != null)
                {
                    this.m_nextStateId = -1;

                    string szTreePath = pNode.GetReferencedTree(pAgent);

                    //to create the task on demand
                    if (this.m_subTree == null || szTreePath != this.m_subTree.GetName())
                    {
                        if (this.m_subTree != null)
                        {
                            Workspace.Instance.DestroyBehaviorTreeTask(this.m_subTree, pAgent);
                        }

                        this.m_subTree = Workspace.Instance.CreateBehaviorTreeTask(szTreePath);
                        pNode.SetTaskParams(pAgent, this.m_subTree);
                    }
                    else if (this.m_subTree != null)
                    {
                        this.m_subTree.reset(pAgent);
                    }

                    pNode.SetTaskParams(pAgent, this.m_subTree);

                    return true;
                }

                return false;
            }

            protected override void onexit(Agent pAgent, EBTStatus s)
            {
#if BEHAVIAC_USE_HTN
                Debug.Check(this.currentState != null);
                this.currentState.Pop();
#endif
                base.onexit(pAgent, s);
            }

            protected override EBTStatus update(Agent pAgent, EBTStatus childStatus)
            {
                ReferencedBehavior pNode = this.GetNode() as ReferencedBehavior;
                Debug.Check(pNode != null);

                if (pNode != null)
                {
#if !BEHAVIAC_RELEASE
                    pAgent.m_debug_count++;

                    if (pAgent.m_debug_count > 20)
                    {
                        Debug.LogWarning(string.Format("{0} might be in a recurrsive inter calling of trees\n", pAgent.GetName()));
                        Debug.Check(false);
                    }
#endif

                    EBTStatus result = this.m_subTree.exec(pAgent);

                    bool bTransitioned = State.UpdateTransitions(pAgent, pNode, pNode.m_transitions, ref this.m_nextStateId, result);

                    if (bTransitioned)
                    {
                        if (result == EBTStatus.BT_RUNNING)
                        {
                            //subtree not exited, but it will transition to other states
                            this.m_subTree.abort(pAgent);
                        }

                        result = EBTStatus.BT_SUCCESS;
                    }

                    return result;
                }

                return EBTStatus.BT_INVALID;
            }
        }
    }
}

File: src/behaviortree/nodes/composites/referencebehavior.cpp
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// Tencent is pleased to support the open source community by making behaviac available.
//
// Copyright (C) 2015-2017 THL A29 Limited, a Tencent company. All rights reserved.
//
// Licensed under the BSD 3-Clause License (the "License"); you may not use this file except in compliance with
// the License. You may obtain a copy of the License at http://opensource.org/licenses/BSD-3-Clause
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is
// distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and limitations under the License.
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////

#include "behaviac/common/base.h"
#include "behaviac/behaviortree/nodes/composites/referencebehavior.h"
#include "behaviac/agent/agent.h"

#include "behaviac/behaviortree/nodes/actions/action.h"

#include "behaviac/htn/planner.h"
#include "behaviac/htn/plannertask.h"
#include "behaviac/htn/task.h"
#include "behaviac/fsm/state.h"
#include "behaviac/fsm/transitioncondition.h"
#include "behaviac/behaviortree/nodes/conditions/condition.h"
#include "behaviac/common/meta.h"

namespace behaviac {
    ReferencedBehavior::ReferencedBehavior() : m_taskNode(0), m_referencedBehaviorPath(0), m_taskMethod(0), m_transitions(0)
    {}

    ReferencedBehavior::~ReferencedBehavior() {
        BEHAVIAC_DELETE(m_referencedBehaviorPath);
        BEHAVIAC_DELETE(m_taskMethod);
        BEHAVIAC_DELETE this->m_transitions;
    }

    void ReferencedBehavior::load(int version, const char* agentType, const properties_t& properties) {
        super::load(version, agentType, properties);

        for (propertie_const_iterator_t it = properties.begin(); it != properties.end(); ++it) {
            const property_t& p = (*it);

            if (StringUtils::StringEqual(p.name, "ReferenceBehavior")) {
                if (StringUtils::IsValidString(p.value)) {
                    const char* pParenthesis = strchr(p.value, '(');

                    if (pParenthesis == 0) {
                        behaviac::string typeName;
                        this->m_referencedBehaviorPath = AgentMeta::ParseProperty(p.value);
                    } else {
                        this->m_referencedBehaviorPath = AgentMeta::ParseMethod(p.value);
                    }

                    const char* szTreePath = this->GetReferencedTree(0);

                    //conservatively make it true
                    bool bHasEvents = true;

                    if (!StringUtils::IsNullOrEmpty(szTreePath)) {
						if (Config::PreloadBehaviors()) {
							//it has a const tree path, so as to load the tree and check if that tree has events
							BehaviorTree* behaviorTree = Workspace::GetInstance()->LoadBehaviorTree(szTreePath);
							BEHAVIAC_ASSERT(behaviorTree);

							if (behaviorTree) {
								bHasEvents = behaviorTree->HasEvents();
							}
						}

						this->m_bHasEvents |= bHasEvents;
					}
                }
			}
			else if (StringUtils::StringEqual(p.name, "Task")) {
                BEHAVIAC_ASSERT(!StringUtils::IsNullOrEmpty(p.value));
                //BEHAVIAC_ASSERT(m is CTaskMethod);

                this->m_taskMethod = AgentMeta::ParseMethod(p.value);
            } else {
                //BEHAVIAC_ASSERT(0, "unrecognised property %s", p.name);
            }
        }
    }

#if BEHAVIAC_USE_HTN

    bool ReferencedBehavior::decompose(BehaviorNode* node, PlannerTaskComplex* seqTask, int depth, Planner* planner) {
        ReferencedBehavior* taskSubTree = (ReferencedBehavior*)node;// as ReferencedBehavior;
        bool bOk = false;
        BEHAVIAC_ASSERT(taskSubTree != 0);
        int depth2 = planner->GetAgent()->GetVariables()->Depth();
        BEHAVIAC_UNUSED_VAR(depth2);
        {
            AgentState::AgentStateScope scopedState(planner->GetAgent()->GetVariables()->Push(false));

            Agent* pAgent = planner->GetAgent();
            const char* szTreePath = taskSubTree->GetReferencedTree(pAgent);
            BehaviorTreeTask* subTreeTask = Workspace::GetInstance()->CreateBehaviorTreeTask(szTreePath);

            //planner.agent.Variables.Log(planner.agent, true);
            taskSubTree->SetTaskParams(planner->GetAgent(), subTreeTask);

            Task* task = taskSubTree->RootTaskNode(planner->GetAgent());

            if (task != 0) {
                planner->LogPlanReferenceTreeEnter(planner->GetAgent(), taskSubTree);

                //const BehaviorNode* tree = task->GetParent();
                //tree->InstantiatePars(planner->GetAgent());
                BehaviorTreeTask* oldCurrentTreeTask = pAgent->m_excutingTreeTask;
                pAgent->m_excutingTreeTask = subTreeTask;

                PlannerTask* childTask = planner->decomposeNode(task, depth);
                pAgent->m_excutingTreeTask = oldCurrentTreeTask;

                if (childTask != 0) {
                    PlannerTaskReference* subTreeRef = (PlannerTaskReference*)seqTask;
                    subTreeRef->SetSubTreeTask(subTreeTask);

                    seqTask->AddChild(childTask);
                    bOk = true;
                }

                planner->LogPlanReferenceTreeExit(planner->GetAgent(), taskSubTree);
                BEHAVIAC_ASSERT(true);
            }
        }

        BEHAVIAC_ASSERT(planner->GetAgent()->GetVariables()->Depth() == depth2);
        return bOk;
    }
#endif

    void ReferencedBehavior::Attach(BehaviorNode* pAttachment, bool bIsPrecondition, bool bIsEffector, bool bIsTransition) {
        if (bIsTransition) {
            BEHAVIAC_ASSERT(!bIsEffector && !bIsPrecondition);

            if (this->m_transitions == 0) {
                this->m_transitions = BEHAVIAC_NEW behaviac::vector<Transition*>();
            }

            BEHAVIAC_ASSERT(Transition::DynamicCast(pAttachment) != 0);
            Transition* pTransition = (Transition*)pAttachment;
            this->m_transitions->push_back(pTransition);

            return;
        }

        BEHAVIAC_ASSERT(bIsTransition == false);
        super::Attach(pAttachment, bIsPrecondition, bIsEffector, bIsTransition);
    }

    bool ReferencedBehavior::IsValid(Agent* pAgent, BehaviorTask* pTask) const {
        if (!ReferencedBehavior::DynamicCast(pTask->GetNode())) {
            return false;
        }

        return super::IsValid(pAgent, pTask);
    }

    BehaviorTask* ReferencedBehavior::createTask() const {
        ReferencedBehaviorTask* pTask = BEHAVIAC_NEW ReferencedBehaviorTask();

        return pTask;
    }

    void ReferencedBehavior::SetTaskParams(Agent* pAgent, BehaviorTreeTask* treeTask) {
        if (this->m_taskMethod != 0) {
            this->m_taskMethod->SetTaskParams(pAgent, treeTask);
        }
    }

    Task* ReferencedBehavior::RootTaskNode(Agent* pAgent) {
        if (this->m_taskNode == 0) {
            BehaviorTree* bt = Workspace::GetInstance()->LoadBehaviorTree(this->GetReferencedTree(pAgent));

            if (bt != 0 && bt->GetChildrenCount() == 1) {
                BehaviorNode* root = (BehaviorNode*)bt->GetChild(0);
                this->m_taskNode = (Task*)root;
            }
        }

        return this->m_taskNode;
    }

    const char* ReferencedBehavior::GetReferencedTree(const Agent* pAgent) const {
        BEHAVIAC_ASSERT(this->m_referencedBehaviorPath);

        if (this->m_referencedBehaviorPath) {
            const char* str = (const char*)m_referencedBehaviorPath->GetValue(pAgent, false, behaviac::GetClassTypeNumberId<const char*>());

            if (str == NULL) {
                return NULL;
            } else {
                return str;// ->c_str();
            }
        }

        return NULL;
    }

	ReferencedBehaviorTask::ReferencedBehaviorTask() : SingeChildTask(), m_nextStateId(-1), m_subTree(NULL) {
    }

    ReferencedBehaviorTask::~ReferencedBehaviorTask() {
		if (this->m_subTree)
		{
			behaviac::Workspace::GetInstance()->DestroyBehaviorTreeTask(this->m_subTree, NULL);
			this->m_subTree = NULL;
		}
    }

    void ReferencedBehaviorTask::copyto(BehaviorTask* target) const {
        super::copyto(target);

        // BEHAVIAC_ASSERT(ReferencedBehaviorTask::DynamicCast(target));
        // ReferencedBehaviorTask* ttask = (ReferencedBehaviorTask*)target;
    }

    void ReferencedBehaviorTask::save(IIONode* node) const {
        super::save(node);
    }

    void ReferencedBehaviorTask::load(IIONode* node) {
        super::load(node);
    }

    void ReferencedBehaviorTask::Init(const BehaviorNode* node) {
        super::Init(node);
    }

    bool ReferencedBehaviorTask::onevent(Agent* pAgent, const char* eventName, behaviac::map<uint32_t, IInstantiatedVariable*>* eventParams) {
        if (this->m_status == BT_RUNNING && this->m_node->HasEvents()) {
            BEHAVIAC_ASSERT(this->m_subTree);

            if (!this->m_subTree->onevent(pAgent, eventName, eventParams)) {
                return false;
            }
        }

        return true;
    }

    bool ReferencedBehaviorTask::onenter(Agent* pAgent) {
        BEHAVIAC_UNUSED_VAR(pAgent);
        BEHAVIAC_ASSERT(ReferencedBehavior::DynamicCast(this->m_node) != 0);
        ReferencedBehavior* pNode = (ReferencedBehavior*)this->m_node;

        BEHAVIAC_ASSERT(pNode != 0);

        this->m_nextStateId = -1;

        const char* szTreePath = pNode->GetReferencedTree(pAgent);

		//to create the task on demand
		if (szTreePath && (this->m_subTree == NULL || !StringUtils::Compare(szTreePath, this->m_subTree->GetName().c_str())))
		{
			if (this->m_subTree)
			{
				behaviac::Workspace::GetInstance()->DestroyBehaviorTreeTask(this->m_subTree, pAgent);
			}

			this->m_subTree = Workspace::GetInstance()->CreateBehaviorTreeTask(szTreePath);
			pNode->SetTaskParams(pAgent, this->m_subTree);
		}
		else if (this->m_subTree)
		{
			this->m_subTree->reset(pAgent);
		}

        return true;
    }

    void ReferencedBehaviorTask::onexit(Agent* pAgent, EBTStatus s) {
#if BEHAVIAC_USE_HTN
        BEHAVIAC_ASSERT(this->m_currentState != NULL);
        this->m_currentState->Pop();
#endif
        BEHAVIAC_UNUSED_VAR(pAgent);
        BEHAVIAC_UNUSED_VAR(s);
    }

    int ReferencedBehaviorTask::GetNextStateId() const {
        return this->m_nextStateId;
    }

    bool ReferencedBehaviorTask::CheckPreconditions(const Agent* pAgent, bool bIsAlive) const {
#if BEHAVIAC_USE_HTN
        ReferencedBehaviorTask* pThis = const_cast<ReferencedBehaviorTask*>(this);
        pThis->m_currentState = const_cast<Agent*>(pAgent)->GetVariables()->Push(false);
        BEHAVIAC_ASSERT(m_currentState != NULL);
#endif//

        bool bOk = BehaviorTask::CheckPreconditions(pAgent, bIsAlive);
#if BEHAVIAC_USE_HTN

        if (!bOk) {
            this->m_currentState->Pop();
            pThis->m_currentState = NULL;
        }

#endif//
        return bOk;
    }

    EBTStatus ReferencedBehaviorTask::update(Agent* pAgent, EBTStatus childStatus) {
        BEHAVIAC_UNUSED_VAR(childStatus);
        BEHAVIAC_ASSERT(ReferencedBehavior::DynamicCast(this->GetNode()));
        const ReferencedBehavior* pNode = (const ReferencedBehavior*)this->m_node;
        BEHAVIAC_ASSERT(pNode);

#if !BEHAVIAC_RELEASE
        pAgent->m_debug_count++;

        if (pAgent->m_debug_count > 20) {
            BEHAVIAC_LOGWARNING("%s might be in a recurrsive inter calling of trees\n", pAgent->GetName().c_str());
            BEHAVIAC_ASSERT(false);
        }

#endif//#if !BEHAVIAC_RELEASE

        EBTStatus result = this->m_subTree->exec(pAgent);

        bool bTransitioned = State::UpdateTransitions(pAgent, pNode, pNode->m_transitions, this->m_nextStateId, result);

        if (bTransitioned) {
            if (result == BT_RUNNING) {
                //subtree not exited, but it will transition to other states
                this->m_subTree->abort(pAgent);
            }

            result = BT_SUCCESS;
        }

        return result;
    }
}//namespace behaviac
